forward_warp/forward_warp.py

```diff
@@ -0,0 +1,191 @@
+    """
+
+    # Flip (x, y) to (y, x) for flow, to avoid further confusion.
+    flow = tf.reverse(flow, axis=[-1])
+
+    # Get target indices along with corresponding values to add.
+    indices, splat_values = get_translated_pixels(features, flow)
+
+    # Mask out out-of-bounds splat values
+    height, width, channels = tf.shape(features)[1], tf.shape(features)[2], tf.shape(features)[3]
+    greater_than_zero = tf.greater_equal(indices, 0)
+    greater_than_zero = tf.cast(tf.reduce_all(greater_than_zero, axis=-1), tf.float32)
+    less_than_height = tf.cast(tf.less(indices[..., 0], height), tf.float32)
+    less_than_width = tf.cast(tf.less(indices[..., 1], width), tf.float32)
+    within_bounds = greater_than_zero * less_than_height * less_than_width
+    splat_values *= tf.expand_dims(within_bounds, axis=-1)
+
+    # Clip indices and flatten.
+    y_indices = tf.clip_by_value(indices[..., 0], 0, height-1)
+    x_indices = tf.clip_by_value(indices[..., 1], 0, width-1)
+    indices = tf.stack([y_indices, x_indices], axis=-1)
+
+    # Scatter into output.
+    all_warped = []
+
+    def _scatter(elms):
+        splat_values, indices, features = elms
+        warped = tf.scatter_nd(indices, splat_values, tf.shape(features))
+        all_warped.append(warped)
+        summed = tf.add_n(all_warped)
+
+        # map_fn requires us to return the same number of things as arguments (nested structure must match).
+        # See: https://stackoverflow.com/questions/47984876/tensorflow-tf-map-fn-parameters
+        return summed, summed, summed
+
+    warped, _, _ = tf.map_fn(_scatter, (splat_values, indices, features), back_prop=True)
+    return warped
+
+
+def get_translated_pixels(features, translations):
+    """
+    :param features: A Tensor. Of shape [batch_size, H, W, C].
+    :param translations: A Tensor. Translations in image pixel units, of shape [batch_size, H, W, 2].
+    :return: indices: Tensor of shape [batch_size, num_indices, 2]. The indices to target.
+             values: Tensor of shape [batch_size, num_indices, C]. The values to put at the corresponding indices.
+    """
+
+    # Get translated mesh-grid.
+    fshape = tf.shape(features)
+    batch_size, height, width, channels = fshape[0], fshape[1], fshape[2], fshape[3]
+    x_1d, y_1d = tf.range(0, width), tf.range(0, height)
+    x_2d, y_2d = tf.meshgrid(x_1d, y_1d)
+    indices_2d = tf.cast(tf.stack([y_2d, x_2d], axis=-1), tf.float32)
+    translated_indices = translations + indices_2d
+
+    # Get splat corners.
+    ceiled = tf.cast(tf.ceil(translated_indices), tf.int32)
+    floored = tf.cast(tf.floor(translated_indices), tf.int32)
+    tl_indices = floored
+    tr_indices = tf.stack([floored[..., 0], ceiled[..., 1]], axis=-1)
+    br_indices = ceiled
+    bl_indices = tf.stack([ceiled[..., 0], floored[..., 1]], axis=-1)
+
+    # If both x and y have integer coordinates (ceil and floor are the same),
```

Comment:
The values are all 0 as you say anyways, so the end-result is easier to achieve the way you wrote it. The only reason why all this weird code exists is to help the flow sub-gradients -- the flow will only "feel" a gradient from the indices that it splatted to, even if the splat value is 0 (specifically, only 1 of the x or y partial derivatives are non-zero when the splat is at integer coordinates). 

So non-zero flow gradients exist:
1.  At 4 indices when translated_index are floats.
2.  At 4 indices (a cross centered on translated_index) when both x, y are integers. 5 splat indices are needed still since the center needs a value.
3.  At 6 indices when only one of x, y are integers.

However, I forgot to handle case 3 properly. I think that the cleanest way to implement this properly is to just splat to 9 locations, so that we are guaranteed to handle all 3 cases.

Reply:
After looking at the spatial transformer network bilinear sampling implementations (which have this same issue), and seeing how they decided to partially ignore it, I think it is safe to just do as you said. I have simplified the implementation.